Use reentrant mode when invoking prims, delete global prim_fake_mode

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84310
* #84096
* __->__ #84090

Maybe I should be using the meta_impl instead of the prim_impl, but it's not terribly clear why, since the prim impl will be better tested and should work under the re-entrant FakeTensorMode.

Fixes https://github.com/pytorch/pytorch/issues/78613 in the process

eellison committed:
Use reentrant mode when invoking prims, delete global prim_fake_mode

[ghstack-poisoned]

eellison committed:
Update on "Use reentrant mode when invoking prims, delete global prim_fake_mode"


Maybe I should be using the meta_impl instead of the prim_impl, but it's not terribly clear why, since the prim impl will be better tested and should work under the re-entrant FakeTensorMode.

Fixes https://github.com/pytorch/pytorch/issues/78613 in the process

[ghstack-poisoned]

eellison committed:
Update on "Use reentrant mode when invoking prims, delete global prim_fake_mode"


Maybe I should be using the meta_impl instead of the prim_impl, but it's not terribly clear why, since the prim impl will be better tested and should work under the re-entrant FakeTensorMode.

Fixes https://github.com/pytorch/pytorch/issues/78613 in the process

[ghstack-poisoned]

bdhirsh commented on torch/_prims/__init__.py:
```diff
@@ -259,22 +238,13 @@ def TensorMeta(
+    if any(
+        isinstance(inp, torch.SymIntNode) for inp in itertools.chain(shape, strides)
+    ):
+        return torch.empty(shape, dtype, device=device)
```
Should this just unconditionally call `empty_strided` once it has symint support?

I can try to update it once https://github.com/pytorch/pytorch/pull/81238/files lands (I'm updating it today)

samdow commented on test/test_ops.py:
```diff
@@ -167,7 +167,9 @@ def test_numpy_ref(self, device, dtype, op):
     @ops(python_ref_db)
     def test_python_ref_meta(self, device, dtype, op):
-        mode = torch._prims.get_prim_fake_mode()
+        mode = FakeTensorMode()
+        with mode:
```
nit: slight preference for `with FakeTensorMode() as mode:` here and L394

eellison committed:
Update on "Use reentrant mode when invoking prims, delete global prim_fake_mode"


Maybe I should be using the meta_impl instead of the prim_impl, but it's not terribly clear why, since the prim impl will be better tested and should work under the re-entrant FakeTensorMode.

Fixes https://github.com/pytorch/pytorch/issues/78613 in the process

[ghstack-poisoned]

eellison commented:
I'm running into an issue with:
`prims.cos` registers a decomposition for `aten.cos` to call into `prims.cos`. `prims.cos` internally invokes `aten.cos`... So you have an infinite loop. I did a somewhat temporary change here just to try to get things to pass that will invoke the meta kernel for `prims.cos` instead of the aten kernel if we're in the middle of running `prims.cos`. Open to suggestions on other changes here. 

I'm running into an error with the SymInt tests. 
```
> torch.empty(shape, dtype=dtype, device=device)
RuntimeError: SymIntArrayRef expected to contain only concrete integers
```

eellison commented on torch/_prims/__init__.py:
```diff
@@ -259,22 +238,13 @@ def TensorMeta(
+    if any(
+        isinstance(inp, torch.SymIntNode) for inp in itertools.chain(shape, strides)
+    ):
+        return torch.empty(shape, dtype=dtype, device=device)
```
hmm, this is failing now..

eellison replied:
when I invoke this with `torch.empty([2, 3], dtype=dtype, device=device)` here a FakeTensor is returned successfully but `torch.ops.aten.empty.SymInt` with SymInt inputs fails with `RuntimeError: SymIntArrayRef expected to contain only concrete integers

cc @horace, @Krovatkin

eellison committed:
Update on "Use reentrant mode when invoking prims, delete global prim_fake_mode"


Maybe I should be using the meta_impl instead of the prim_impl, but it's not terribly clear why, since the prim impl will be better tested and should work under the re-entrant FakeTensorMode.

Fixes https://github.com/pytorch/pytorch/issues/78613 in the process

[ghstack-poisoned]

eellison committed:
Update on "Use reentrant mode when invoking prims, delete global prim_fake_mode"


Maybe I should be using the meta_impl instead of the prim_impl, but it's not terribly clear why, since the prim impl will be better tested and should work under the re-entrant FakeTensorMode.

Fixes https://github.com/pytorch/pytorch/issues/78613 in the process

[ghstack-poisoned]

eellison committed:
Update on "Use reentrant mode when invoking prims, delete global prim_fake_mode"


Maybe I should be using the meta_impl instead of the prim_impl, but it's not terribly clear why, since the prim impl will be better tested and should work under the re-entrant FakeTensorMode.

Fixes https://github.com/pytorch/pytorch/issues/78613 in the process

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison committed:
Update on "Use reentrant mode when invoking prims, delete global prim_fake_mode"


Maybe I should be using the meta_impl instead of the prim_impl, but it's not terribly clear why, since the prim impl will be better tested and should work under the re-entrant FakeTensorMode.

Fixes https://github.com/pytorch/pytorch/issues/78613 in the process

[ghstack-poisoned]

eellison committed:
Update on "Use reentrant mode when invoking prims, delete global prim_fake_mode"


Maybe I should be using the meta_impl instead of the prim_impl, but it's not terribly clear why, since the prim impl will be better tested and should work under the re-entrant FakeTensorMode.

Fixes https://github.com/pytorch/pytorch/issues/78613 in the process

[ghstack-poisoned]

eellison committed:
Update on "Use reentrant mode when invoking prims, delete global prim_fake_mode"


Maybe I should be using the meta_impl instead of the prim_impl, but it's not terribly clear why, since the prim impl will be better tested and should work under the re-entrant FakeTensorMode.

Fixes https://github.com/pytorch/pytorch/issues/78613 in the process

[ghstack-poisoned]

eellison committed:
Update on "Use reentrant mode when invoking prims, delete global prim_fake_mode"


Maybe I should be using the meta_impl instead of the prim_impl, but it's not terribly clear why, since the prim impl will be better tested and should work under the re-entrant FakeTensorMode.

Fixes https://github.com/pytorch/pytorch/issues/78613 in the process

[ghstack-poisoned]

eellison commented:
@pytorchmergebot help

eellison commented:
@pytorchbot --help

eellison commented:
@pytorchbot merge --green

eellison commented:
cc @malfet  @seemethere this is failing with 
```
ios-12-5-1-arm64-coreml - Unauthorized
ios-12-5-1-arm64-custom-ops - Unauthorized
[macos-12-py3-x86-64-build](https://circleci.com/gh/pytorch/pytorch/17066636?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link) - Success
[macos-12-py3-x86-64-test-2-2-default](https://circleci.com/gh/pytorch/pytorch/17066711?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link) - Success
[macos-12-py3-arm64-build](https://circleci.com/gh/pytorch/pytorch/17066635?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link) - Success
ios-12-5-1-arm64-metal - Unauthorized
ios-12-5-1-arm64 - Unauthorized
``` 

infra error ?

malfet commented:
@eellison yes, please ignore

malfet commented:
@pytorchbot merge -f "CircleCI failures are spurious"
